### PR TITLE
LPD-37516 Prevent more than one ThemeCSSCET from being applied at the control panel level at the same time

### PR DESCRIPTION
**Pull request by thiago-buarqque**
@drewbrokke what do you think?

### CODE REVIEW

File: modules/apps/client-extension/client-extension-service/src/main/java/com/liferay/client/extension/internal/events/ClientExtensionsServicePreAction.java [hunk 1]
```diff
@@ -54,6 +56,9 @@ public void run(
 
 		Layout layout = themeDisplay.getLayout();
 
+		_setThemeDisplayCSSURLs(
+			httpServletRequest, _getThemeCSSCET(layout), themeDisplay);
```
Comment (drewbrokke) [inline modules/apps/client-extension/client-extension-service/src/main/java/com/liferay/client/extension/internal/events/ClientExtensionsServicePreAction.java:60]: Inline

File: modules/apps/client-extension/client-extension-service/src/main/java/com/liferay/client/extension/internal/events/ClientExtensionsServicePreAction.java [hunk 2]
```diff
@@ -175,6 +193,12 @@ private String _getFaviconURL(Layout layout) {
 	}
 
 	private ThemeCSSCET _getThemeCSSCET(Layout layout) {
+		if (layout.isTypeControlPanel() &&
+			FeatureFlagManagerUtil.isEnabled("LPD-34650")) {
```
Comment (drewbrokke) [inline modules/apps/client-extension/client-extension-service/src/main/java/com/liferay/client/extension/internal/events/ClientExtensionsServicePreAction.java:197]: Move the feature flag check inside the `_getControlPanelThemeCSSCET` method and return null

File: modules/apps/client-extension/client-extension-test/src/testIntegration/java/com/liferay/client/extension/type/internal/configuration/test/CETConfigurationFactoryTest.java [hunk 3]
```diff
@@ -109,6 +124,69 @@ public void testActivate() throws Exception {
...
+				).build());
+
+		try {
+			Layout layout = _getControlPanelLayout(company.getCompanyId());
+
+			Assert.assertNotNull(
+				_clientExtensionEntryRelLocalService.
+					fetchClientExtensionEntryRel(
+						_portal.getClassNameId(Layout.class), layout.getPlid(),
+						ClientExtensionEntryConstants.TYPE_THEME_CSS));
```
Comment (drewbrokke) [inline modules/apps/client-extension/client-extension-test/src/testIntegration/java/com/liferay/client/extension/type/internal/configuration/test/CETConfigurationFactoryTest.java:166]: Assert on the log when adding a second one

File: modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/configuration/CETConfigurationFactory.java [hunk 4]
```diff
@@ -74,25 +108,114 @@ protected void modified(Map<String, Object> properties) throws Exception {
...
+					_clientExtensionEntryRelLocalService.
+						deleteClientExtensionEntryRels(
+							companyId, _cet.getExternalReferenceCode());
+
+					_addThemeCSSClientExtensionEntryRel(companyId);
+				}
 			});
 	}
 
+	private void _addThemeCSSClientExtensionEntryRel(Long companyId)
```
Comment (drewbrokke) [inline modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/configuration/CETConfigurationFactory.java:125]: Does this structure make things clearer?

```
fetch by erc

if not null,
	return

add

count

if count > 1
	log error
```

**Review by drewbrokke — CHANGES_REQUESTED** [review]

File: modules/apps/client-extension/client-extension-test/src/testIntegration/java/com/liferay/client/extension/type/internal/configuration/test/CETConfigurationFactoryTest.java [hunk 5]
```diff
@@ -109,6 +126,118 @@ public void testActivate() throws Exception {
...
+
+			name = RandomTestUtil.randomString();
+
+			try (LogCapture logCapture = LoggerTestUtil.configureLog4JLogger(
+					"com.liferay.client.extension.type.internal." +
+						"configuration.CETConfigurationFactory",
+					LoggerTestUtil.ERROR)) {
```
Comment (drewbrokke) [inline modules/apps/client-extension/client-extension-test/src/testIntegration/java/com/liferay/client/extension/type/internal/configuration/test/CETConfigurationFactoryTest.java:175]: This try-with-resources statement can just open the whole block, no need to nest `try` statements.

File: modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/configuration/CETConfigurationFactory.java [hunk 6]
```diff
@@ -74,25 +108,111 @@ protected void modified(Map<String, Object> properties) throws Exception {
...
+			_portal.getClassNameId(Layout.class), layout.getPlid(),
+			_cet.getExternalReferenceCode(), _cet.getType(), StringPool.BLANK,
+			new ServiceContext());
+
+		List<ClientExtensionEntryRel> clientExtensionEntryRels =
+			_clientExtensionEntryRelLocalService.getClientExtensionEntryRels(
+				_portal.getClassNameId(Layout.class), layout.getPlid(),
+				ClientExtensionEntryConstants.TYPE_THEME_CSS);
+
+		if (clientExtensionEntryRels.size() > 1) {
```
Comment (drewbrokke) [inline modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/configuration/CETConfigurationFactory.java:160]: This works, but `_clientExtensionEntryRelLocalService.getClientExtensionEntryRelsCount` might marginally more efficient.

**Review by drewbrokke — CHANGES_REQUESTED** [review]

File: modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/configuration/CETConfigurationFactory.java [hunk 7]
```diff
@@ -74,25 +108,112 @@ protected void modified(Map<String, Object> properties) throws Exception {
...
+		int clientExtensionEntryRelsCount =
+			_clientExtensionEntryRelLocalService.
+				getClientExtensionEntryRelsCount(
+					_portal.getClassNameId(Layout.class), layout.getPlid(),
+					ClientExtensionEntryConstants.TYPE_THEME_CSS);
+
+		if (clientExtensionEntryRelsCount > 1) {
+			_log.error(
+				"Only one theme CSS client extension can be applied to the " +
+					"control panel at the same time");
```
[the remark this answers is not in the document]
Comment (thiago-buarqque) [reply modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/configuration/CETConfigurationFactory.java:164]: Let's bring this to the platform experience repo, will ask Evan's opinion as well

[the remark this answers is not in the document]
Comment (thiago-buarqque) [reply modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/configuration/CETConfigurationFactory.java:164]: But I will do the rewording

**Comment by thiago-buarqque** [comment]
@drewbrokke continues on https://github.com/liferay-platform-experience/liferay-portal/pull/770

**Pull request closed without merging**